### ISSUE TITLE
how to connect to remote clickhouse server

### ISSUE THREAD

**Issue by SE2AI** [mymarilyn/clickhouse-driver#185]
actually i think its a really useful tool, but the documentations are sooo poor.
the introduction provides both client and connector examples, however all of these are toy examples. just as below:
>client = Client('localhost')
>conn = connect('clickhouse://localhost')

how to build connection in real production env is not mentioned, how about remote situations? how to config clickhouse-server? which parameters are needed for client and connector API? 
None of these clearly provided. so i think this project is built from your daily work. but the project has got 0.5k stars, i got confused.

a better choice is https://github.com/ClickHouse/clickhouse-go

**Comment by xzkostyan** [comment]
Did you actually read the docs? https://clickhouse-driver.readthedocs.io

**Comment by vialorn** [comment]
* `host`
* `port`
* `database`
* `user`
* `password`
and etc. - are all described in the [docs](https://clickhouse-driver.readthedocs.io/en/latest/api.html#connection).

**Comment by SE2AI** [comment]
> Did you actually read the docs? https://clickhouse-driver.readthedocs.io

it will be convenient if an example attached.

**Comment by SE2AI** [comment]
> * `host`
> * `port`
> * `database`
> * `user`
> * `password`
>   and etc. - are all described in the [docs](https://clickhouse-driver.readthedocs.io/en/latest/api.html#connection).

i post two examples here:

* show tables with Client
```
from clickhouse_driver import Client
client = Client(host='xx.xx.xx.xx', port='9000', database='data_base_name', user='default', password='123456')
ret_show = client.execute('SHOW TABLES')
print(ret_show)
```

* show tables with connect
```
from clickhouse_driver import connect
conn = connect('clickhouse://default:123456@xx.xx.xx.xx:9000/data_base_name')
cursor = conn.cursor()
cursor.execute('SHOW TABLES')
res = cursor.fetchall()
print(res)
```

@vialorn thx! i have some other questions:
1. what are the differences between _Client_ and _Connection_ in the doc? (I found _Client_ is easily to use)
2. what are the differences between API and DB API2? 
3. which API is better to use for what kind of situations?

**Comment by vialorn** [comment]
1. `Client` uses `Connection` under the hood. There is no need to use `Connection` directly. I agree the docs is not as descriptive as newbies may desire it.
2. DB API2 is a standard Python API for DB interaction. It may be convenient for one to stick to it instead on learning DBMS-specific API.
3. Perhaps if one develops a program that should support different DBMS - DB API will be of use. But with all ClickHouse behaviour specific, I hardly can imagine such a program. ClickHouse does not support a good part of classic MVCC RDBMS features implemented in DB API and vice versa. ClickHouse has many features that cannot be crammed into DB API specifications, neither at the moment nor soon.

**Comment by SE2AI** [comment]
@vialorn Thanks a lot for your time and patience! The answer above is really helpful. 

Recently I conducted some research on clickhouse, and also found it's different from MySQL, especially the inner mechanisms.
According to answer 3, DB API2 is convenient but it may not support all kinds of DMLs. For some simple operations, such as INSERT/UPDATE/DELETE，DB API2 is a better choice, right?

* The last question :)
For INSERT/UPDATE/DELETE operation, neither `client.execute` nor `cursor.execute/cursor.fetchall` will return `[ ]`. However, based on this result, we cannot be sure about whether clickhouse process the query correctly or not, 
So, how do we know the behavior of clickhouse? If the operation failed in clickhouse, will the API throw some exceptions? 
I found `exception clickhouse_driver.dbapi.OperationalError` in online-docs, is it designed for this purpose?

**Comment by vialorn** [comment]
@SE2AI,
The first thing you should fathom about ClickHouse: there are NO `UPDATE` / `DELETE` in a way like they are in MVCC RDBMS (Oracle, MSSQL, PostgreSQL, MySQL). ClickHouse is all about `append-only` data. There are fast unblocked `INSERT` and usually super-fast aggregation and lookup `SELECT`.
The `UPDATE` and `DELETE` are async. One sends them to ClickHouse and server immediately returns `OK`. But the real work will be performed in background in future. May be in seconds, maybe in hours.

If
* a query is incorrect
* an error occurs while performing a sync query (`INSERT` / `SELECT`)

then `clickhouse_driver` throws an exception, of course.

**Comment by SE2AI** [comment]
@vialorn ok I see, when searching other issues, I found a way to making `UPDARTE `and `DELETE `synchronous, from https://clickhouse.tech/docs/en/operations/settings/settings/#mutations_sync.

it says:
> mutations_sync 
> Allows to execute ALTER TABLE ... UPDATE|DELETE queries (mutations) synchronously.
> 
> Possible values:
> 
> 0 - Mutations execute asynchronously.
> 1 - The query waits for all mutations to complete on the current server.
> 2 - The query waits for all mutations to complete on all replicas (if they exist).
> Default value: 0.

I have not tried it, perhaps it has some implied side-effects, such as the `UPDATE` is too slow, so the `INSERT` has to wait.
Do you have any idea why `UPDATE` will be performed in hours?
If there is a batch of `UPDATE` at the same time (e.g. lots of sources send update commands), what the clickhouse performance will be? Will the server get stuck?

**Comment by vialorn** [comment]
In the case of row-oriented RDBMS, `UPDATE` or `DELETE` means just an update of a small database "page".
In the case of column-based DBMS with compressed column data - `UPDATE` or `DELETE` means read a whole `part` and write this `part` back in a modified state. The bigger `part` is - the more resources this process consumes. In addition, these `mutation` operations have lower priority than `INSERT`s, small part merges and `SELECT`s.
As of a synchronous option - this just leads to a client to wait for `mutation` to complete. This way an issuer may be sure that on-disk data is modified but the speed of operation is still unpleasantly slow.

**Comment by SE2AI** [comment]
Hi @vialorn , I notice that clickhouse doc also mentions when `SELECTS`/`INSERT` simultaneously, or `SELECT`/`UPDATE` simultaneously，the table will not be locked. But `INSERT`/`UPDATE` will block each other:
https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/mergetree/#concurrent-data-access

Do you know whether `PARTITION` could solve the `INSERT`/`UPDATE`-lock problem? 
By setting `PARTITION` during create the table, the records will be inserted into different partitions according to some specific columns. Assuming that when clickhouse perform `INSERT` in **partition_1** and `UPDATE` in **partition_2**, will these two action block each other? Or operations on different partitions are performed independently?

**Comment by vialorn** [comment]
@SE2AI, I guess you better address ClickHouse (not clickhouse_driver) related questions to Telegram channel:
https://t.me/clickhouse_en, or
https://t.me/clickhouse_ru.
I have no experience with `en` version, but on `ru` I've got a lot of help.
